### PR TITLE
Min SDK 21 and other version updates

### PR DESCRIPTION
**Pull request by mattprecious**

### CODE REVIEW

File: swirl-sample/build.gradle [hunk 1]
```diff
@@ -9,12 +9,13 @@ apply plugin: 'com.android.application'
...
-  compile 'com.jakewharton:butterknife:7.0.1'
+  implementation 'com.jakewharton:butterknife:8.8.1'
+  annotationProcessor 'com.jakewharton:butterknife-compiler:8.8.1'
 }
 
 android {
-  compileSdkVersion 23
-  buildToolsVersion "23.0.3"
+  compileSdkVersion 27
+  buildToolsVersion "27.0.3"
```
Comment (JakeWharton) [inline swirl-sample/build.gradle]: You can delete this. AGP will pick an appropriate version automatically.

File: swirl/build.gradle [hunk 2]
```diff
@@ -1,17 +1,18 @@
...
   }
 }
 
 apply plugin: 'com.android.library'
 
 android {
-  compileSdkVersion 23
-  buildToolsVersion '23.0.3'
+  compileSdkVersion 27
+  buildToolsVersion '27.0.3'
```
Comment (JakeWharton) [inline swirl/build.gradle]: Same as above

File: swirl/build.gradle [hunk 3]
```diff
@@ -1,17 +1,18 @@
...
 
   defaultConfig {
     versionName VERSION_NAME
+    vectorDrawables.useSupportLibrary = true
```
Comment (JakeWharton) [inline swirl/build.gradle]: There's no minSdk defined. Is it actually min 21 or are you lowering the functionality to work on 21? If it's min 21 then you shouldn't actually need this.

File: swirl/src/main/java/com/mattprecious/swirl/SwirlView.java [hunk 4]
```diff
@@ -3,14 +3,15 @@
...
 import android.graphics.drawable.Drawable;
 import android.os.Build;
 import android.support.annotation.DrawableRes;
+import android.support.graphics.drawable.AnimatedVectorDrawableCompat;
+import android.support.graphics.drawable.VectorDrawableCompat;
 import android.util.AttributeSet;
 import android.widget.ImageView;
 
-@TargetApi(Build.VERSION_CODES.M)
+@TargetApi(Build.VERSION_CODES.LOLLIPOP)
```
Comment (JakeWharton) [inline swirl/src/main/java/com/mattprecious/swirl/SwirlView.java:15]: ...or this (with min defined)

File: swirl/src/main/java/com/mattprecious/swirl/SwirlView.java [hunk 5]
```diff
@@ -49,13 +50,19 @@ public void setState(State state, boolean animate) {
...
+        icon = AnimatedVectorDrawableCompat.create(getContext(), resId);
+      }
+      if (icon == null) {
+        icon = VectorDrawableCompat.create(getResources(), resId, getContext().getTheme());
+      }
       setImageDrawable(icon);
 
-      if (icon instanceof AnimatedVectorDrawable) {
-        ((AnimatedVectorDrawable) icon).start();
+      if (icon instanceof AnimatedVectorDrawableCompat) {
```
Comment (JakeWharton) [inline swirl/src/main/java/com/mattprecious/swirl/SwirlView.java]: Switch to `Animatable`?

**Review by JakeWharton — APPROVED** [review]

File: swirl/build.gradle (hunk 3 above)
Comment (mattprecious) [reply swirl/build.gradle]: Lowering to work with 21. If I specify an actual min in the build file it'll bump the min of projects that pull this in, right?

Comment (JakeWharton) [reply swirl/build.gradle]: Yes. You should at least specify 14 (which is what the support libs use). Without one you get 1 which adds some weird permission for compatibility reasons pre-3.

**Comment by NightlyNexus** [comment]
Why the need for the support library's VectorDrawableCompat if the constructor ensures only Lollipop+ usage?

**Comment by JakeWharton** [comment]
Native vectors are only used on 23 or 24+, and that might change in the
future. Basically providing backported features and critical bug fixes.

**Comment by NightlyNexus** [comment]
Cool, thanks for the info! 👍

**Pull request merged**